src/main/java/org/dcsa/tnt/service/impl/EventSubscriptionServiceImpl.java

```diff
@@ -17,20 +23,74 @@
+    private final NotificationServiceConfig notificationServiceConfig;
+    private final NotificationSignatureHandler notificationSignatureHandler;
+
 
     @Override
     public EventSubscriptionRepository getRepository() {
         return eventSubscriptionRepository;
     }
 
     @Override
-    protected Mono<EventSubscription> preSaveHook(EventSubscription eventSubscription) {
+    protected Mono<EventSubscription> preCreateHook(EventSubscription eventSubscription) {
+        byte[] secret = eventSubscription.getSecret();
+        SignatureMethod signatureMethod;
+        if (eventSubscription.getSignatureMethod() == null) {
+            signatureMethod = notificationServiceConfig.getDefaultSignatureMethod();
+            eventSubscription.setSignatureMethod(signatureMethod);
+        } else {
+            signatureMethod = eventSubscription.getSignatureMethod();
+        }
+        if (secret == null || secret.length < 1) {
+            return Mono.error(new CreateException("Please provide a non-empty \"secret\" attribute"));
+        }
+        if (signatureMethod.getMinKeyLength() == signatureMethod.getMaxKeyLength()) {
+            if (secret.length != signatureMethod.getMinKeyLength()) {
+                if (secret.length  > 1 && secret.length - 1 == signatureMethod.getMinKeyLength()
+                        && Character.isSpaceChar(((int)secret[secret.length - 1]) & 0xff)) {
+                    return Mono.error(new CreateException("The secret provided in the \"secret\" attribute must be exactly "
+                            + signatureMethod.getMinKeyLength() + " bytes long (when deserialized).  Did you"
+                            + " accidentally include a trailing newline/space in the secret?"));
+                }
+                return Mono.error(new CreateException("The secret provided in the \"secret\" attribute must be exactly "
+                        + signatureMethod.getMinKeyLength() + " bytes long (when deserialized)"));
+            }
+        } else {
+            if (secret.length < signatureMethod.getMinKeyLength()) {
+                return Mono.error(new CreateException("The secret provided in the \"secret\" attribute must be at least "
+                        + signatureMethod.getMinKeyLength() + " bytes long (when deserialized)"));
+            }
+            if (secret.length > signatureMethod.getMaxKeyLength()) {
+                return Mono.error(new CreateException("The secret provided in the \"secret\" attribute must be at most "
+                        + signatureMethod.getMaxKeyLength() + " bytes long (when deserialized)"));
+            }
+        }
+        return checkCallbackUrl(eventSubscription);
+    }
+
+    @Override
+    protected Mono<EventSubscription> preUpdateHook(EventSubscription original, EventSubscription update) {
+        return checkCallbackUrl(update);
+    }
+
+    protected Mono<EventSubscription> checkCallbackUrl(EventSubscription eventSubscription) {
```

Comment:
Shouldn't we do a HEAD request here to ensure that the URL exists

Reply:
True - but this is also not a new bug (the old spec also had that rule). :)

Reply:
Fair enough - just remember to add it later ;-)